remove redundant to_dtype in Fused Schedular Nodes

zhuhaozhe opened the pull request:
Fix https://github.com/pytorch/pytorch/issues/115260.
This issue is triggered by `FusedSchedularNodes` cases.
We always store `lowp buffer` to `store_cache` then load `lowp buffer` from `store_cache` and `convert it to float` before `compute ops`.
Now we will generate a `{key: to(float32)_expr, value: the float32 cse var before to_dtype and store}` in `cse.cache`.
Then the `to_dtype(float32)` after `load` will hit this cache and not generate a new var with cast codes.


Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #118365



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames

zhuhaozhe committed:
remove rebundant to_dtype in Fused Schedular Nodes

[ghstack-poisoned]

zhuhaozhe committed:
Update on "remove rebundant to_dtype in Fused Schedular Nodes"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

zhuhaozhe commented:
@pytorchbot rebase

zhuhaozhe committed:
Update on "remove rebundant to_dtype in Fused Schedular Nodes"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

zhuhaozhe committed:
Update on "remove rebundant to_dtype in Fused Schedular Nodes"

Fix https://github.com/pytorch/pytorch/issues/115260.



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

zhuhaozhe committed:
Update on "remove rebundant to_dtype in Fused Schedular Nodes"

Fix https://github.com/pytorch/pytorch/issues/115260.



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

zhuhaozhe committed:
Update on "remove rebundant to_dtype in Fused Schedular Nodes"

Fix https://github.com/pytorch/pytorch/issues/115260.
This issue is triggered by `FusedSchedularNodes` cases.
We always store `lowp buffer` to `store_cache` then load `lowp buffer` from `store_cache` and `convert it to float` before `compute ops`.
Now we will also store `float buffer` so we can load `float buffer` from `store_cache` and directly do compute on it.





cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

zhuhaozhe committed:
Update on "remove rebundant to_dtype in Fused Schedular Nodes"

Fix https://github.com/pytorch/pytorch/issues/115260.
This issue is triggered by `FusedSchedularNodes` cases.
We always store `lowp buffer` to `store_cache` then load `lowp buffer` from `store_cache` and `convert it to float` before `compute ops`.
Now we will also store `float buffer` so we can load `float buffer` from `store_cache` and directly do compute on it.





cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

zhuhaozhe committed:
Update on "remove rebundant to_dtype in Fused Schedular Nodes"

Fix https://github.com/pytorch/pytorch/issues/115260.
This issue is triggered by `FusedSchedularNodes` cases.
We always store `lowp buffer` to `store_cache` then load `lowp buffer` from `store_cache` and `convert it to float` before `compute ops`.
Now we will generate a `{key: to(float32)_expr, value: the float32 cse var before to_dtype and store}` in `cse.cache`.
Then the `to_dtype(float32)` after `load` will hit this cache and not generate a new var with cast codes.





cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

jgong5 commented on torch/_inductor/codegen/cpp.py:
```diff
@@ -1326,6 +1336,10 @@ def to_dtype(x, dtype, src_dtype=None):
         ], f"{__name__} does not support {dtype}"
         node: torch.fx.Node = V.interpreter.current_node
         assert node and isinstance(node, torch.fx.Node)
+        if node.target == "store":
+            assert dtype == torch.float
+            assert src_dtype in DTYPE_LOWP_FP
+            return f"cvt_lowp_fp_to_fp32<{DTYPE_TO_CPP[src_dtype]}>({x})"
```
This looks a bit hacky. Perhaps you can factor out a method to return the line for such a conversion, used by both `to_dtype` and `store`?

zhuhaozhe replied:
Removed this part in to_dtypes.
Instead, just add a function to generate `to_fp32_expr`

jgong5 commented on torch/_inductor/codegen/cpp.py:
```diff
@@ -1483,6 +1497,32 @@ def masked(self, mask):
             self._load_mask = prior
 
+    def cache_fp32_cse_var_before_store(self, var_to_store):
+        def find_fp32_var(var, cache):
```
add comment on what the function does.

zhuhaozhe replied:
Added

zhuhaozhe committed:
Update on "remove rebundant to_dtype in Fused Schedular Nodes"

Fix https://github.com/pytorch/pytorch/issues/115260.
This issue is triggered by `FusedSchedularNodes` cases.
We always store `lowp buffer` to `store_cache` then load `lowp buffer` from `store_cache` and `convert it to float` before `compute ops`.
Now we will generate a `{key: to(float32)_expr, value: the float32 cse var before to_dtype and store}` in `cse.cache`.
Then the `to_dtype(float32)` after `load` will hit this cache and not generate a new var with cast codes.





cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

zhuhaozhe commented on test/inductor/test_cpu_repro.py:
```diff
@@ -2943,6 +2945,35 @@ def fn(x):
         self.common(fn, (x,))
         assert metrics.generated_cpp_vec_kernel_count == 1
 
+    def test_no_rebundant_to_dtypes_between_fused_scheduler_node(self):
```
Fixed thanks

jgong5 commented on torch/_inductor/codegen/cpp.py:
```diff
@@ -1527,6 +1580,7 @@ def load(self, name: str, index: sympy.Expr):
     def store(self, name, index, value, mode=None):
         assert "buf" in name
         var = self.args.output(name)
+        self.cache_fp32_cse_var_before_store(value)
```
I guess we only need to invoke this method when `value` is bf16/fp16. Should we add a check before calling it?

zhuhaozhe replied:
Added lowp data type check

zhuhaozhe commented on torch/_inductor/codegen/cpp.py:
```diff
@@ -969,6 +987,7 @@ def sign(x):
         code.writeline(f"auto {left} = {x} > 0 ? {scalar_one} : {scalar_zero};")
         code.writeline(f"auto {right} = {x} < 0 ? {scalar_one} : {scalar_zero};")
         code.writeline(f"auto {result} = {left} - {right};")
+        V.kernel.cse.cache[f"auto {result} = {left} - {right};"] = result
```
this cse value is not generated by `cse.generate` so it will not in the `cse cache`.
so if the graph is
```
tmp1 = sign(tmp0)
tmp2 = to_bfloat16(tmp1)
store(tmp2)
```
we will not be able to find `tmp1` in `cse.cache`.

jgong5 replied:
should we make `sign` use `cse.generate` instead?

zhuhaozhe replied:
Sure, I updated it.

zhuhaozhe commented on torch/_inductor/codegen/cpp.py:
```diff
@@ -280,6 +280,14 @@ def is_to_lowp_dtype(expr):
     return None
 
 
+def to_fp32_expr(lowp_var, src_dtype, kernel):
```
thanks for the adviced, changed function name.

zhuhaozhe committed:
Update on "remove rebundant to_dtype in Fused Schedular Nodes"

Fix https://github.com/pytorch/pytorch/issues/115260.
This issue is triggered by `FusedSchedularNodes` cases.
We always store `lowp buffer` to `store_cache` then load `lowp buffer` from `store_cache` and `convert it to float` before `compute ops`.
Now we will generate a `{key: to(float32)_expr, value: the float32 cse var before to_dtype and store}` in `cse.cache`.
Then the `to_dtype(float32)` after `load` will hit this cache and not generate a new var with cast codes.





cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

zhuhaozhe committed:
Update on "remove rebundant to_dtype in Fused Schedular Nodes"

Fix https://github.com/pytorch/pytorch/issues/115260.
This issue is triggered by `FusedSchedularNodes` cases.
We always store `lowp buffer` to `store_cache` then load `lowp buffer` from `store_cache` and `convert it to float` before `compute ops`.
Now we will generate a `{key: to(float32)_expr, value: the float32 cse var before to_dtype and store}` in `cse.cache`.
Then the `to_dtype(float32)` after `load` will hit this cache and not generate a new var with cast codes.





cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

zhuhaozhe committed:
Update on "remove rebundant to_dtype in Fused Schedular Nodes"

Fix https://github.com/pytorch/pytorch/issues/115260.
This issue is triggered by `FusedSchedularNodes` cases.
We always store `lowp buffer` to `store_cache` then load `lowp buffer` from `store_cache` and `convert it to float` before `compute ops`.
Now we will generate a `{key: to(float32)_expr, value: the float32 cse var before to_dtype and store}` in `cse.cache`.
Then the `to_dtype(float32)` after `load` will hit this cache and not generate a new var with cast codes.





cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

jgong5 commented on torch/_inductor/codegen/cpp.py:
```diff
@@ -963,14 +981,12 @@ def sign(x):
         # auto tmp5 = tmp4 < 0 ? -1 : 1;
         left = V.kernel.cse.newvar()
         right = V.kernel.cse.newvar()
-        result = V.kernel.cse.newvar()
         scalar_zero = f"decltype({x})(0)"
         scalar_one = f"decltype({x})(1)"
         code.writeline(f"auto {left} = {x} > 0 ? {scalar_one} : {scalar_zero};")
         code.writeline(f"auto {right} = {x} < 0 ? {scalar_one} : {scalar_zero};")
-        code.writeline(f"auto {result} = {left} - {right};")
         V.kernel.compute.splice(code)
-        return result
+        return f"{left} - {right}"
```
Consider to use the lambda function to make sure cse cache takes effect.
```python
code.writeline("[&]()")
with code.indent():
    code.writeline(f"auto left = {x} > 0 ? {scalar_one} : {scalar_zero};")
    code.writeline(f"auto right = {x} < 0 ? {scalar_one} : {scalar_zero};")
    code.writeline(f"return left - right;")
return code
```

zhuhaozhe replied:
Updated, thanks

jgong5 commented on torch/_inductor/codegen/cpp.py:
```diff
@@ -1483,6 +1497,45 @@ def masked(self, mask):
+    def cache_fp32_cse_var_before_store(self, var_to_store):
+        """
+        https://github.com/pytorch/pytorch/issues/115260
+        For FusedSchedulerNode[node1, node2], the node2 might need to load node1's result
+        the node1's result might be lowp data type and node2 might need to cast node1's lowp dtype
+        to fp32 to do following computes. Thus we add a pair
+        {to_fp32_expr, node1's fp32 store cse var} in cse cache here to avoid the "to_dtype"
+        after node2's "load".
```
```suggestion
        For FusedSchedulerNode[node1, node2], the node2 loads what node1 stores and the buffer is
        in low-precision floating point data type. When the output of node1 also serves as the output of the
        kernel, the result of nodes would be different from the case when output of node1 is not the output
        of the kernel (where we don't need to insert `to_dtype` for legalization). To address the problem, on
        storing the lowp node1 output, we also add the inverse dtype conversion to high precision data type
        to the cse cache.
        
        Example (pseudo code):
            node1_output = ...
            node1_output_lowp = to_dtype(node1_output, dtype=torch.bfloat16)
            store(buf, node1_output_lowp)
            node2_input_lowp = load(buf)
            node2_input = to_dtype(node2_input_lowp, dtype=torch.float)
            
        Without cse cache trick:
            node1_output = ...
            node1_output_lowp = to_dtype(node1_output, dtype=torch.bfloat16)
            store(buf, node1_output_lowp)
            node2_input_lowp = node_output_lowp # hit store cache
            node2_input = to_dtype(node2_input_lowp, dtype=torch.float)
        
        With cse cache trick:
            node1_output = ...
            node1_output_lowp = to_dtype(node1_output, dtype=torch.bfloat16)
            # also add `to_dtype(node1_input_lowp, dtype=torch.float)` -> `node1_output` to cse cache
            store(buf, node1_output_lowp)
            node2_input_lowp = node_output_lowp # hit store cache
            node2_input = node1_output # hit cse cache
```

jgong5 commented on torch/_inductor/codegen/cpp.py:
```diff
@@ -1483,6 +1497,45 @@ def masked(self, mask):
         finally:
             self._load_mask = prior
 
+    def cache_fp32_cse_var_before_store(self, var_to_store):
```
```suggestion
    def cache_fp32_cse_var_before_lowp_store(self, var_to_store):
```

zhuhaozhe committed:
Update on "remove rebundant to_dtype in Fused Schedular Nodes"

Fix https://github.com/pytorch/pytorch/issues/115260.
This issue is triggered by `FusedSchedularNodes` cases.
We always store `lowp buffer` to `store_cache` then load `lowp buffer` from `store_cache` and `convert it to float` before `compute ops`.
Now we will generate a `{key: to(float32)_expr, value: the float32 cse var before to_dtype and store}` in `cse.cache`.
Then the `to_dtype(float32)` after `load` will hit this cache and not generate a new var with cast codes.





cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

jgong5 commented on torch/_inductor/codegen/cpp.py:
```diff
@@ -1314,19 +1319,21 @@ def where(a, b, c):
     def sign(x):
         code = BracesBuffer()
-        # auto tmp5 = tmp4 < 0 ? -1 : 1;
+        """
+        auto left = decltype(tmp1)::blendv(decltype(tmp1)(0), decltype(tmp1)(1), decltype(tmp1)(0) < tmp1);
+        auto right = decltype(tmp1)::blendv(decltype(tmp1)(0), decltype(tmp1)(1), tmp1 < decltype(tmp1)(0));
+        return left - right;
+        """
```
why adding this comment. The code looks simple enough to know what code it would generate...

zhuhaozhe replied:
Removed.

zhuhaozhe committed:
Update on "remove rebundant to_dtype in Fused Schedular Nodes"

Fix https://github.com/pytorch/pytorch/issues/115260.
This issue is triggered by `FusedSchedularNodes` cases.
We always store `lowp buffer` to `store_cache` then load `lowp buffer` from `store_cache` and `convert it to float` before `compute ops`.
Now we will generate a `{key: to(float32)_expr, value: the float32 cse var before to_dtype and store}` in `cse.cache`.
Then the `to_dtype(float32)` after `load` will hit this cache and not generate a new var with cast codes.





cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

jansel commented:
Is this ready for review?  Still marked as draft.

jgong5 commented:
> Is this ready for review? Still marked as draft.

@jansel Yes, please review.

zhuhaozhe commented:
@pytorchbot merge

zhuhaozhe committed:
Update on "remove redundant to_dtype in Fused Schedular Nodes"

Fix https://github.com/pytorch/pytorch/issues/115260.
This issue is triggered by `FusedSchedularNodes` cases.
We always store `lowp buffer` to `store_cache` then load `lowp buffer` from `store_cache` and `convert it to float` before `compute ops`.
Now we will generate a `{key: to(float32)_expr, value: the float32 cse var before to_dtype and store}` in `cse.cache`.
Then the `to_dtype(float32)` after `load` will hit this cache and not generate a new var with cast codes.





cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]

zhuhaozhe commented:
@pytorchbot merge

zhuhaozhe commented:
@pytorchbot rebase

zhuhaozhe commented:
@pytorchbot merge
